File: stripe/src/main/java/com/stripe/android/IssuingCardEphemeralKey.java
package com.stripe.android;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

final class IssuingCardEphemeralKey extends EphemeralKey {
    public static final Parcelable.Creator<IssuingCardEphemeralKey> CREATOR
            = new Parcelable.Creator<IssuingCardEphemeralKey>() {

        @Override
        public IssuingCardEphemeralKey createFromParcel(Parcel in) {
            return new IssuingCardEphemeralKey(in);
        }

        @Override
        public IssuingCardEphemeralKey[] newArray(int size) {
            return new IssuingCardEphemeralKey[size];
        }
    };

    private IssuingCardEphemeralKey(@NonNull Parcel in) {
        super(in);
    }

    private IssuingCardEphemeralKey(long created, @NonNull String customerId, long expires,
                                    @NonNull String id, boolean liveMode, @NonNull String object,
                                    @NonNull String secret, @NonNull String type) {
        super(created, customerId, expires, id, liveMode, object, secret, type);
    }

    @NonNull
    String getIssuingCardId() {
        return mObjectId;
    }

    static final class Factory extends EphemeralKey.Factory<IssuingCardEphemeralKey> {
        @NonNull
        @Override
        IssuingCardEphemeralKey create(long created, @NonNull String objectId, long expires,
                                       @NonNull String id, boolean liveMode, @NonNull String object,
                                       @NonNull String secret, @NonNull String type) {
            return new IssuingCardEphemeralKey(created, objectId, expires, id, liveMode, object,
                    secret, type);
        }
    }
}

File: stripe/src/main/java/com/stripe/android/IssuingCardEphemeralKey.kt
package com.stripe.android

import android.os.Parcel
import android.os.Parcelable

internal class IssuingCardEphemeralKey : EphemeralKey {

    val issuingCardId: String
        get() = objectId

    private constructor(parcel: Parcel) : super(parcel)

    private constructor(
        created: Long,
        customerId: String,
        expires: Long,
        id: String,
        liveMode: Boolean,
        objectType: String,
        secret: String,
        type: String
    ) : super(created, customerId, expires, id, liveMode, objectType, secret, type)

    internal class Factory : EphemeralKey.Factory<IssuingCardEphemeralKey>() {
        override fun create(
            created: Long,
            objectId: String,
            expires: Long,
            id: String,
            liveMode: Boolean,
            objectType: String,
            secret: String,
            type: String
        ): IssuingCardEphemeralKey {
            return IssuingCardEphemeralKey(
                created, objectId, expires, id, liveMode, objectType, secret, type
            )
        }
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<IssuingCardEphemeralKey> =
            object : Parcelable.Creator<IssuingCardEphemeralKey> {
                override fun createFromParcel(parcel: Parcel): IssuingCardEphemeralKey {
                    return IssuingCardEphemeralKey(parcel)
                }

                override fun newArray(size: Int): Array<IssuingCardEphemeralKey?> {
                    return arrayOfNulls(size)
                }
            }
    }
}
